### PR TITLE
[Bugfix] Fixed epoch level schedulers not being called when val_check_interval < 1.0

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Currently, epoch level learning rate schedulers are updated after each validation epoch, when it should be after each training epoch. This can be seen by adding `val_check_interval=0.5` to many of our scheduler test, and they will fail since the learning rate gets updated twice per epoch. This PR fixes it.

Edit: actually the bug seems to be that epoch level learning rate schedulers never seems to be called when `val_check_interval!=1`, so the test fails because the learning rate is unaltered.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
mind ad chlog

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -750,10 +750,6 @@ def run_evaluation(self, max_batches=None, on_epoch=False):
         # hook
         self.evaluation_loop.on_evaluation_epoch_end()
 
-        # update epoch-level lr_schedulers
-        if on_epoch:
-            self.optimizer_connector.update_learning_rates(interval='epoch')
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:755]: the reason I added this to update learning_rate inorder to have the correct state for lr_schedulers since they are save in `on_validation_end` hook in the next line. `on_epoch=True` only when the train loop is finished and should not be a problem when val_check_interval < 1. since in such a case val loop is called within the training loop up there where on_epoch=False (default) so it shouldn't be updating this learning rate.

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -561,17 +561,18 @@ def run_training_epoch(self):
 
         should_check_val = self.should_check_val_fx(batch_idx, is_last_batch, on_epoch=True)
         if should_check_val:
-            self.trainer.run_evaluation(on_epoch=True)
+            self.trainer.run_evaluation()
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/training_loop.py:564]: okay I see the problem here. what we can do there is have a flag (bool) or something that will indicate whether val_loop is called or not inside the training loop (val_check_interval < 1.), and if it's true then we can just update the lr before the val loop outside training loop is triggered to solve both the conditions here. 

```python
if should_val_check or val_loop_called:
    self.optimizer_connector.update_learning_rates(interval='epoch')

if should_val_check:
    self.trainer.run_evaluation()
```

will this work?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (SkafteNicki) [reply pytorch_lightning/trainer/trainer.py:755]: okay, now I better understand why it is there. I originally hesitated to remove it because I thought that there must have been a good reason why you put it there in the first place :]

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (SkafteNicki) [reply pytorch_lightning/trainer/training_loop.py:566]: learning rates needs to be updated after `run_evaluation` since some schedulers like `
ReduceLROnPlateau` can be condition on specific metrics calculated in the validation loop

Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:566]: then the checkpoint won't be getting the correct state since `on_validation_end` for checkpoint callback is called inside validation_loop. I am assuming that before on_validation_end call in the val loop, everything logged inside `LightningModule` hooks or log-supported callback hooks must be aggregated correctly. If they aren't then that would be a problem with other callbacks like `ModelCheckpoint`, `EarlyStopping` too right?

@tchaton can you confirm whether everything logged is aggregated before [this call](https://github.com/PyTorchLightning/pytorch-lightning/blob/0b271474e5cc35a40009559d6c4488ab3b02723e/pytorch_lightning/trainer/trainer.py#L758) or not?

File: tests/checkpointing/test_model_checkpoint.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ def configure_optimizers(self):
         assert mc_specific_data['current_score'] == score
 
         lr_scheduler_specific_data = chk['lr_schedulers'][0]
-        assert lr_scheduler_specific_data['_step_count'] == epoch + 2
+        assert lr_scheduler_specific_data['_step_count'] == epoch + 1
```
Comment (rohitgr7) [inline tests/checkpointing/test_model_checkpoint.py:134]: actually it should be +2 since epoch here is zero indexed (0, 1, 2) and default_value for `_step_count` is 1. so for total_epochs=1 it should be 1(default) + (epoch(0) + 1) = 2

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:566]: hmm.. I think 
```
if val_loop_called:
    self.optimizer_connector.update_learning_rates(interval='epoch')
if should_val_check:
    self.trainer.run_evaluation(on_epoch=True)
```
should be enough and rest as it is. I hope didn't miss any edge case this time.

Although there is still one more edge case here. If user sets `val_check_interval=0.3` or something then metric used might not be the correct one that will be used to make decision inside `ReductLRonPleateau`, but I'd say that is more of the logic/implementation error from user's side.

Comment (SkafteNicki) [reply pytorch_lightning/trainer/training_loop.py:566]: alright I try to revert back, and see if it fixes the problem

File: tests/checkpointing/test_model_checkpoint.py (hunk 3 above)
Comment (SkafteNicki) [reply tests/checkpointing/test_model_checkpoint.py]: changed back :]

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py]: oh sorry! missed one more edge case here :sweat_smile: 

```
if val_loop_called and not should_check_val:
```
to handle cases when val_check_interval=0.1, 0.2, 0.25, 0.5

thinking for more edges cases here to avoid any possible failures in future.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py]: > thinking for more edges cases here to avoid any possible failures in future.

Can we write tests for all these corner cases? Better safe than sorry

Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py]: added some tests. found a bug too.

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py]: Is this ready-to-go then?

Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py]: Need to add a similar strong test for reducelronplateau scheduler.

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
Awesome @SkafteNicki @rohitgr7 !

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -558,21 +560,23 @@ def run_training_epoch(self):
...
+        # update epoch level lr_schedulers if no val loop outside train loop is triggered
+        if (val_loop_called and not should_check_val) or should_train_only:
             self.trainer.optimizer_connector.update_learning_rates(interval='epoch')
+
+        if should_train_only:
             self.check_checkpoint_callback(True)
             self.check_early_stopping_callback(True)
 
+        if should_check_val:
+            self.trainer.run_evaluation(on_epoch=True)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:575]: are these checks guaranteed to be mutually exclusive? i'm updating to 1.2.4 and see a test failure with one of my modules where it looks like there's a gap: 
before: `run_evaluation` ran first, allowing the module to run the validation loop. inside validation step/validation epoch end, the module could log metrics. afterward, the checkpoint is force-run. so if the checkpoint configured a metric that was available only during validation, then somehow this still worked. 
by moving the run_evaluation to happen after the force checkpoint saving, we fail when looking up the monitor value

i think the correct fix is dropping the training loop force running checkpoint/early stopping callbacks here. they should be part of the callback.

Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:575]: I believe if validation is disabled or val batches=0 then only we run the force checkpoint else it will call run_evaluation and the the callbacks will be called inside and monitor will be taken care of as expected. Can you paste a small example with your case where it doesn't work?

Comment (ananthsub) [reply pytorch_lightning/trainer/training_loop.py:575]: It's not, see https://github.com/PyTorchLightning/pytorch-lightning/pull/7207 

this PR introduced the change where we call the evaluation loop after we call the checkpoint/early stopping callbacks from training. 

As a result, this check for `should_train_only` is incomplete - it inherently depends on the evaluation loop to populate `num_val_batches` correctly. in `run_evaluation` we set the validation dataloader, but this is too late as the validation dataloader is what's used to determine `should_skip_eval` above
